tests/ConnectionTest.php

```diff
@@ -134,6 +137,114 @@ public function testConnectDispatchEvent(): void
         $conn->connect();
     }
 
+    public function testTransactionBeginDispatchEvent(): void
+    {
+        $listenerMock = $this->createMock(TransactionBeginDispatchEventListener::class);
+        $listenerMock->expects(self::exactly(2))->method('onTransactionBegin');
+
+        $eventManager = new EventManager();
+        $eventManager->addEventListener([Events::onTransactionBegin], $listenerMock);
+        $listener = new class {
+            /** @var Connection|null */
+            private $connection = null;
+
+            public function onTransactionBegin(TransactionBeginEventArgs $args): void
+            {
+                $this->connection = $args->getConnection();
+            }
+
+            public function getConnection(): ?Connection
+            {
+                return $this->connection;
+            }
+        };
+        $eventManager->addEventListener([Events::onTransactionBegin], $listener);
+
+        $driverMock = $this->createMock(Driver::class);
+
+        $conn = new Connection([], $driverMock, new Configuration(), $eventManager);
+        $conn->beginTransaction();
+        self::assertSame($conn, $listener->getConnection());
+        self::assertEquals(1, $listener->getConnection()->getTransactionNestingLevel());
```

Comment:
The transaction nesting level doesn't really look relevant to the logic being tested: the only responsibility of the connection is to call the listener and pass itself as an argument.

If you want to have a more complex scenario covered by a test, I believe it's better to be done in a separate test.

Reply:
Yeah, I guess `self::assertSame($conn, $listener->getConnection());` is enough.

Reply:
It's better if the mock just asserts that it's called with the expected arguments (`$args->getConnection()`), it shouldn't contain any state.

Reply:
Yes, but how do I assert that? Since it's called with `Event` class with wraps the connection. Is that possible with mock?

Reply:
Yeah, you can do something like:
```php
$listener->expects(self::once())
    ->method('onTransactionBegin')
    ->with($this->callback(function (TransactionBeginEventArgs $args) use ($conn) {
        $this->assertSame($conn, $args->getConnection());
    }));
```

Reply:
Yeah, refactored it like that:
```php
        $listenerMock = $this->createMock(TransactionRollBackDispatchEventListener::class);
        $listenerMock
            ->expects(self::exactly(1))
            ->method('onTransactionRollBack')
            ->with(
                self::callback(
                    static function (TransactionRollBackEventArgs $eventArgs) use ($conn): bool {
                        return $eventArgs->getConnection() === $conn;
                    }
                )
            );
```
It can be seen now.